csi/src/dev.rs

```diff
@@ -0,0 +1,175 @@
+//! Definition of traits required for attaching and detaching devices.
+//! Note that (unfortunately) the attach and detach operations are not
+//! quite symmetric. This is because it is not possible to discover all
+//! the information contained in a device URI from the entries in udev.
+//! Hence there are separate traits for attach and detach that each
+//! device type must implement.
+//!
+//! Attaching a device is performed as follows:
+//! ```ignore
+//!     let uri = "iscsi://192.168.0.20:3260/iqn.2019-05.io.openebs:nexus-11111111-0000-0000-0000-000000000000/0";
+//!     let device = Device::parse(uri)?;
+//!     if let Some(path) = device.find().await? {
+//!         // device already attached
+//!     } else {
+//!         // attach the device
+//!         device.attach().await?;
+//!         // wait for it to show up in udev and obtain the path
+//!         let path = Device::wait_for_device(device, 10).await?;
+//!     }
+//! ```
+//!
+//! Detaching a device is performed via:
+//! ```ignore
+//!     let uuid = Uuid::parse_str(&volume_id)?;
+//!     if let Some(device) = Device::lookup(&uuid).await? {
+//!         device.detach().await?;
+//!     }
+//! ```
+
+use std::{convert::TryFrom, time::Duration};
+
+use tokio::time::delay_for;
+use udev::Enumerator;
+use url::Url;
+use uuid::Uuid;
+
+mod iscsi;
+mod nbd;
+mod nvmf;
+mod util;
+
+const NVME_NQN_PREFIX: &str = "nqn.2019-05.io.openebs";
+
+pub use crate::error::DeviceError;
+pub type DeviceName = String;
+
+#[tonic::async_trait]
+pub trait Attach: Sync + Send {
+    async fn attach(&self) -> Result<(), DeviceError>;
```

Comment:
Why not -> Result<deviceName, DeviceError> ? The find (which must be implemented) is implicit for attach to succeed right? As we return from attach only when we found the device?

Reply:
At present the attach() returns immediately without waiting to verify that the device has appeared in udev. The reason for this is that, according to Blaise, Kubernetes may abort (and then retry) the CSI call if they take to long. By being able to call find independently means that the CSI call has more chance of being successful on a retry. Again this should be revisited as part of CAS-376